{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: scikit-learn in /Users/sheanfuphen/repos/Factoromer/.venv/lib/python3.9/site-packages (1.5.2)\n",
      "Requirement already satisfied: numpy>=1.19.5 in /Users/sheanfuphen/repos/Factoromer/.venv/lib/python3.9/site-packages (from scikit-learn) (2.0.2)\n",
      "Requirement already satisfied: scipy>=1.6.0 in /Users/sheanfuphen/repos/Factoromer/.venv/lib/python3.9/site-packages (from scikit-learn) (1.13.1)\n",
      "Requirement already satisfied: joblib>=1.2.0 in /Users/sheanfuphen/repos/Factoromer/.venv/lib/python3.9/site-packages (from scikit-learn) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in /Users/sheanfuphen/repos/Factoromer/.venv/lib/python3.9/site-packages (from scikit-learn) (3.5.0)\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import sys\n",
    "\n",
    "# List of packages to install\n",
    "packages = [\n",
    "    'numpy', 'pandas', 'matplotlib', 'mpl_toolkits.mplot3d', 'os', 're', 'warnings', 'seaborn',\n",
    "    'scikit-learn', 'itertools', 'collections', 'adjustText', 'plotly'\n",
    "]\n",
    "\n",
    "# Function to install a package\n",
    "def install(package):\n",
    "    subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", package])\n",
    "\n",
    "# Loop through the list of packages and install them if not already installed\n",
    "for package in packages:\n",
    "    try:\n",
    "        __import__(package)\n",
    "    except ImportError:\n",
    "        install(package)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#importing and loading necessary packages\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import os\n",
    "import re\n",
    "import warnings\n",
    "import seaborn\n",
    "from sklearn import (metrics, tree)\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import (accuracy_score, auc, confusion_matrix, precision_score,\n",
    "                             recall_score, roc_auc_score, roc_curve, classification_report)\n",
    "from sklearn.model_selection import (cross_val_score, cross_validate, train_test_split, StratifiedShuffleSplit)\n",
    "from sklearn.preprocessing import (OneHotEncoder, label_binarize, StandardScaler)\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.tree import (DecisionTreeClassifier, export_graphviz)\n",
    "from itertools import cycle\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from collections import OrderedDict\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from adjustText import adjust_text\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gl/4s8z78lj6j91bb4dytb1dk7h0000gn/T/ipykernel_60667/3587693066.py:3: DtypeWarning: Columns (44,49,88) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mut_data = pd.read_csv('data_mutations.txt', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>VARIANT_CLASS</th>\n",
       "      <th>all_effects</th>\n",
       "      <th>amino_acid_change</th>\n",
       "      <th>cDNA_Change</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>cdna_change</th>\n",
       "      <th>comments</th>\n",
       "      <th>n_depth</th>\n",
       "      <th>t_depth</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>55242470</td>\n",
       "      <td>55242487</td>\n",
       "      <td>+</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>5159</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>149513271</td>\n",
       "      <td>149513271</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBM10</td>\n",
       "      <td>8241</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>47041565</td>\n",
       "      <td>47041598</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>7578235</td>\n",
       "      <td>7578235</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>7577058</td>\n",
       "      <td>7577058</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208948</th>\n",
       "      <td>PTPRT</td>\n",
       "      <td>0</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>20</td>\n",
       "      <td>41408878</td>\n",
       "      <td>41408878</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208949</th>\n",
       "      <td>FLT4</td>\n",
       "      <td>0</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>180055897</td>\n",
       "      <td>180055897</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208950</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>0</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>76940086</td>\n",
       "      <td>76940086</td>\n",
       "      <td>+</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208951</th>\n",
       "      <td>BTK</td>\n",
       "      <td>0</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>100608310</td>\n",
       "      <td>100608310</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208952</th>\n",
       "      <td>ERG</td>\n",
       "      <td>0</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>21</td>\n",
       "      <td>39764351</td>\n",
       "      <td>39764352</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208953 rows × 123 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  \\\n",
       "0             EGFR            1956  MSKCC     GRCh37          7   \n",
       "1           PDGFRB            5159  MSKCC     GRCh37          5   \n",
       "2            RBM10            8241  MSKCC     GRCh37          X   \n",
       "3             TP53            7157  MSKCC     GRCh37         17   \n",
       "4             TP53            7157  MSKCC     GRCh37         17   \n",
       "...            ...             ...    ...        ...        ...   \n",
       "208948       PTPRT               0  MSKCC     GRCh37         20   \n",
       "208949        FLT4               0  MSKCC     GRCh37          5   \n",
       "208950        ATRX               0  MSKCC     GRCh37          X   \n",
       "208951         BTK               0  MSKCC     GRCh37          X   \n",
       "208952         ERG               0  MSKCC     GRCh37         21   \n",
       "\n",
       "        Start_Position  End_Position Strand              Consequence  \\\n",
       "0             55242470      55242487      +         inframe_deletion   \n",
       "1            149513271     149513271      +         missense_variant   \n",
       "2             47041565      47041598      +       frameshift_variant   \n",
       "3              7578235       7578235      +         missense_variant   \n",
       "4              7577058       7577058      +              stop_gained   \n",
       "...                ...           ...    ...                      ...   \n",
       "208948        41408878      41408878      +         missense_variant   \n",
       "208949       180055897     180055897      +         missense_variant   \n",
       "208950        76940086      76940086      +  splice_acceptor_variant   \n",
       "208951       100608310     100608310      +         missense_variant   \n",
       "208952        39764351      39764352      +       frameshift_variant   \n",
       "\n",
       "       Variant_Classification  ... VARIANT_CLASS all_effects  \\\n",
       "0                In_Frame_Del  ...           NaN         NaN   \n",
       "1           Missense_Mutation  ...           NaN         NaN   \n",
       "2             Frame_Shift_Del  ...           NaN         NaN   \n",
       "3           Missense_Mutation  ...           NaN         NaN   \n",
       "4           Nonsense_Mutation  ...           NaN         NaN   \n",
       "...                       ...  ...           ...         ...   \n",
       "208948      Missense_Mutation  ...           NaN         NaN   \n",
       "208949      Missense_Mutation  ...           NaN         NaN   \n",
       "208950            Splice_Site  ...           NaN         NaN   \n",
       "208951      Missense_Mutation  ...           NaN         NaN   \n",
       "208952        Frame_Shift_Del  ...           NaN         NaN   \n",
       "\n",
       "       amino_acid_change cDNA_Change cDNA_position  cdna_change comments  \\\n",
       "0                    NaN         NaN           NaN          NaN      NaN   \n",
       "1                    NaN         NaN           NaN          NaN      NaN   \n",
       "2                    NaN         NaN           NaN          NaN      NaN   \n",
       "3                    NaN         NaN           NaN          NaN      NaN   \n",
       "4                    NaN         NaN           NaN          NaN      NaN   \n",
       "...                  ...         ...           ...          ...      ...   \n",
       "208948               NaN         NaN           NaN          NaN      NaN   \n",
       "208949               NaN         NaN           NaN          NaN      NaN   \n",
       "208950               NaN         NaN           NaN          NaN      NaN   \n",
       "208951               NaN         NaN           NaN          NaN      NaN   \n",
       "208952               NaN         NaN           NaN          NaN      NaN   \n",
       "\n",
       "        n_depth  t_depth  transcript  \n",
       "0           NaN      NaN         NaN  \n",
       "1           NaN      NaN         NaN  \n",
       "2           NaN      NaN         NaN  \n",
       "3           NaN      NaN         NaN  \n",
       "4           NaN      NaN         NaN  \n",
       "...         ...      ...         ...  \n",
       "208948      NaN      NaN         NaN  \n",
       "208949      NaN      NaN         NaN  \n",
       "208950      NaN      NaN         NaN  \n",
       "208951      NaN      NaN         NaN  \n",
       "208952      NaN      NaN         NaN  \n",
       "\n",
       "[208953 rows x 123 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the TSV file saved as a TXT file into a pandas DataFrame\n",
    "clinical_sample = pd.read_csv('data_clinical_sample.txt', sep='\\t')\n",
    "mut_data = pd.read_csv('data_mutations.txt', sep='\\t')\n",
    "mut_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GRCh37']\n",
      "['MSKCC']\n",
      "[nan 'NEWRECORD']\n"
     ]
    }
   ],
   "source": [
    "# checking to see if all samples used the same reference seq\n",
    "print(mut_data['NCBI_Build'].unique())\n",
    "print(mut_data['Center'].unique())\n",
    "print(mut_data['IS_NEW'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Mutation_Status</th>\n",
       "      <th>Score</th>\n",
       "      <th>t_ref_count</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>n_ref_count</th>\n",
       "      <th>n_alt_count</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>Codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956</td>\n",
       "      <td>7</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TAAGAGAAGCAACATCTC</td>\n",
       "      <td>TAAGAGAAGCAACATCTC</td>\n",
       "      <td>-</td>\n",
       "      <td>P-0081657-T01-IM7</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>MSK-IMPACT</td>\n",
       "      <td>319</td>\n",
       "      <td>288</td>\n",
       "      <td>281</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000275493.2:c.2240_2257del</td>\n",
       "      <td>p.Leu747_Pro753delinsSer</td>\n",
       "      <td>p.L747_P753delinsS</td>\n",
       "      <td>tTAAGAGAAGCAACATCTCcg/tcg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>5159</td>\n",
       "      <td>5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>P-0081657-T01-IM7</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>MSK-IMPACT</td>\n",
       "      <td>370</td>\n",
       "      <td>85</td>\n",
       "      <td>333</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000261799.4:c.812A&gt;G</td>\n",
       "      <td>p.His271Arg</td>\n",
       "      <td>p.H271R</td>\n",
       "      <td>cAc/cGc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBM10</td>\n",
       "      <td>8241</td>\n",
       "      <td>X</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>ACTACAATGCTCAGAGCCAGCAGTACCTGTACTG</td>\n",
       "      <td>ACTACAATGCTCAGAGCCAGCAGTACCTGTACTG</td>\n",
       "      <td>-</td>\n",
       "      <td>P-0081657-T01-IM7</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>MSK-IMPACT</td>\n",
       "      <td>189</td>\n",
       "      <td>150</td>\n",
       "      <td>404</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000329236.7:c.1556_1589del</td>\n",
       "      <td>p.Tyr519TrpfsTer96</td>\n",
       "      <td>p.Y519Wfs*96</td>\n",
       "      <td>tACTACAATGCTCAGAGCCAGCAGTACCTGTACTGg/tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157</td>\n",
       "      <td>17</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>P-0083825-T01-IM7</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>MSK-IMPACT</td>\n",
       "      <td>187</td>\n",
       "      <td>111</td>\n",
       "      <td>789</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000269305.4:c.614A&gt;G</td>\n",
       "      <td>p.Tyr205Cys</td>\n",
       "      <td>p.Y205C</td>\n",
       "      <td>tAt/tGt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157</td>\n",
       "      <td>17</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P-0083825-T01-IM7</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>MSK-IMPACT</td>\n",
       "      <td>222</td>\n",
       "      <td>125</td>\n",
       "      <td>797</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000269305.4:c.880G&gt;T</td>\n",
       "      <td>p.Glu294Ter</td>\n",
       "      <td>p.E294*</td>\n",
       "      <td>Gag/Tag</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Chromosome         Consequence  \\\n",
       "0        EGFR            1956          7    inframe_deletion   \n",
       "1      PDGFRB            5159          5    missense_variant   \n",
       "2       RBM10            8241          X  frameshift_variant   \n",
       "3        TP53            7157         17    missense_variant   \n",
       "4        TP53            7157         17         stop_gained   \n",
       "\n",
       "  Variant_Classification Variant_Type                    Reference_Allele  \\\n",
       "0           In_Frame_Del          DEL                  TAAGAGAAGCAACATCTC   \n",
       "1      Missense_Mutation          SNP                                   T   \n",
       "2        Frame_Shift_Del          DEL  ACTACAATGCTCAGAGCCAGCAGTACCTGTACTG   \n",
       "3      Missense_Mutation          SNP                                   T   \n",
       "4      Nonsense_Mutation          SNP                                   C   \n",
       "\n",
       "                    Tumor_Seq_Allele1 Tumor_Seq_Allele2 Tumor_Sample_Barcode  \\\n",
       "0                  TAAGAGAAGCAACATCTC                 -    P-0081657-T01-IM7   \n",
       "1                                   T                 C    P-0081657-T01-IM7   \n",
       "2  ACTACAATGCTCAGAGCCAGCAGTACCTGTACTG                 -    P-0081657-T01-IM7   \n",
       "3                                   T                 C    P-0083825-T01-IM7   \n",
       "4                                   C                 A    P-0083825-T01-IM7   \n",
       "\n",
       "  Mutation_Status       Score  t_ref_count  t_alt_count  n_ref_count  \\\n",
       "0         SOMATIC  MSK-IMPACT          319          288          281   \n",
       "1         SOMATIC  MSK-IMPACT          370           85          333   \n",
       "2         SOMATIC  MSK-IMPACT          189          150          404   \n",
       "3         SOMATIC  MSK-IMPACT          187          111          789   \n",
       "4         SOMATIC  MSK-IMPACT          222          125          797   \n",
       "\n",
       "   n_alt_count                             HGVSc                     HGVSp  \\\n",
       "0            0  ENST00000275493.2:c.2240_2257del  p.Leu747_Pro753delinsSer   \n",
       "1            0        ENST00000261799.4:c.812A>G               p.His271Arg   \n",
       "2            0  ENST00000329236.7:c.1556_1589del        p.Tyr519TrpfsTer96   \n",
       "3            0        ENST00000269305.4:c.614A>G               p.Tyr205Cys   \n",
       "4            0        ENST00000269305.4:c.880G>T               p.Glu294Ter   \n",
       "\n",
       "          HGVSp_Short                                   Codons  \n",
       "0  p.L747_P753delinsS                tTAAGAGAAGCAACATCTCcg/tcg  \n",
       "1             p.H271R                                  cAc/cGc  \n",
       "2        p.Y519Wfs*96  tACTACAATGCTCAGAGCCAGCAGTACCTGTACTGg/tg  \n",
       "3             p.Y205C                                  tAt/tGt  \n",
       "4             p.E294*                                  Gag/Tag  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop columns with all NaN values in clinical_sample\n",
    "clinical_sample = clinical_sample.dropna(axis=1, how='all')\n",
    "\n",
    "# Drop columns with all NaN values in mut_data\n",
    "mut_data = mut_data.dropna(axis=1, how='all')\n",
    "\n",
    "# Drop columns if they exist in mut_data\n",
    "columns_to_drop = ['Center', 'NCBI_Build', 'Start_Position', 'End_Position', 'Strand', 'IS_NEW', 'Protein_position', 'Transcript_ID', 'RefSeq', 'Exon_Number', 'dbSNP_RS', 'Validation_Status', 'ALLELE_NUM']\n",
    "mut_data = mut_data.drop(columns=[col for col in columns_to_drop if col in mut_data.columns])\n",
    "\n",
    "mut_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>...</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>n_ref_count</th>\n",
       "      <th>n_alt_count</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>Codons</th>\n",
       "      <th>tumor_vaf</th>\n",
       "      <th>normal_vaf</th>\n",
       "      <th>Hugo_HGVSp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956</td>\n",
       "      <td>7</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TAAGAGAAGCAACATCTC</td>\n",
       "      <td>TAAGAGAAGCAACATCTC</td>\n",
       "      <td>-</td>\n",
       "      <td>P-0081657-T01-IM7</td>\n",
       "      <td>...</td>\n",
       "      <td>288</td>\n",
       "      <td>281</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000275493.2:c.2240_2257del</td>\n",
       "      <td>p.Leu747_Pro753delinsSer</td>\n",
       "      <td>p.L747_P753delinsS</td>\n",
       "      <td>tTAAGAGAAGCAACATCTCcg/tcg</td>\n",
       "      <td>0.474465</td>\n",
       "      <td>0.0</td>\n",
       "      <td>EGFR_p.Leu747_Pro753delinsSer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>5159</td>\n",
       "      <td>5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>P-0081657-T01-IM7</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>333</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000261799.4:c.812A&gt;G</td>\n",
       "      <td>p.His271Arg</td>\n",
       "      <td>p.H271R</td>\n",
       "      <td>cAc/cGc</td>\n",
       "      <td>0.186813</td>\n",
       "      <td>0.0</td>\n",
       "      <td>PDGFRB_p.His271Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBM10</td>\n",
       "      <td>8241</td>\n",
       "      <td>X</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>ACTACAATGCTCAGAGCCAGCAGTACCTGTACTG</td>\n",
       "      <td>ACTACAATGCTCAGAGCCAGCAGTACCTGTACTG</td>\n",
       "      <td>-</td>\n",
       "      <td>P-0081657-T01-IM7</td>\n",
       "      <td>...</td>\n",
       "      <td>150</td>\n",
       "      <td>404</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000329236.7:c.1556_1589del</td>\n",
       "      <td>p.Tyr519TrpfsTer96</td>\n",
       "      <td>p.Y519Wfs*96</td>\n",
       "      <td>tACTACAATGCTCAGAGCCAGCAGTACCTGTACTGg/tg</td>\n",
       "      <td>0.442478</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RBM10_p.Tyr519TrpfsTer96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157</td>\n",
       "      <td>17</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>P-0083825-T01-IM7</td>\n",
       "      <td>...</td>\n",
       "      <td>111</td>\n",
       "      <td>789</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000269305.4:c.614A&gt;G</td>\n",
       "      <td>p.Tyr205Cys</td>\n",
       "      <td>p.Y205C</td>\n",
       "      <td>tAt/tGt</td>\n",
       "      <td>0.372483</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TP53_p.Tyr205Cys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157</td>\n",
       "      <td>17</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P-0083825-T01-IM7</td>\n",
       "      <td>...</td>\n",
       "      <td>125</td>\n",
       "      <td>797</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000269305.4:c.880G&gt;T</td>\n",
       "      <td>p.Glu294Ter</td>\n",
       "      <td>p.E294*</td>\n",
       "      <td>Gag/Tag</td>\n",
       "      <td>0.360231</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TP53_p.Glu294Ter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Chromosome         Consequence  \\\n",
       "0        EGFR            1956          7    inframe_deletion   \n",
       "1      PDGFRB            5159          5    missense_variant   \n",
       "2       RBM10            8241          X  frameshift_variant   \n",
       "3        TP53            7157         17    missense_variant   \n",
       "4        TP53            7157         17         stop_gained   \n",
       "\n",
       "  Variant_Classification Variant_Type                    Reference_Allele  \\\n",
       "0           In_Frame_Del          DEL                  TAAGAGAAGCAACATCTC   \n",
       "1      Missense_Mutation          SNP                                   T   \n",
       "2        Frame_Shift_Del          DEL  ACTACAATGCTCAGAGCCAGCAGTACCTGTACTG   \n",
       "3      Missense_Mutation          SNP                                   T   \n",
       "4      Nonsense_Mutation          SNP                                   C   \n",
       "\n",
       "                    Tumor_Seq_Allele1 Tumor_Seq_Allele2 Tumor_Sample_Barcode  \\\n",
       "0                  TAAGAGAAGCAACATCTC                 -    P-0081657-T01-IM7   \n",
       "1                                   T                 C    P-0081657-T01-IM7   \n",
       "2  ACTACAATGCTCAGAGCCAGCAGTACCTGTACTG                 -    P-0081657-T01-IM7   \n",
       "3                                   T                 C    P-0083825-T01-IM7   \n",
       "4                                   C                 A    P-0083825-T01-IM7   \n",
       "\n",
       "   ... t_alt_count n_ref_count  n_alt_count                             HGVSc  \\\n",
       "0  ...         288         281            0  ENST00000275493.2:c.2240_2257del   \n",
       "1  ...          85         333            0        ENST00000261799.4:c.812A>G   \n",
       "2  ...         150         404            0  ENST00000329236.7:c.1556_1589del   \n",
       "3  ...         111         789            0        ENST00000269305.4:c.614A>G   \n",
       "4  ...         125         797            0        ENST00000269305.4:c.880G>T   \n",
       "\n",
       "                      HGVSp         HGVSp_Short  \\\n",
       "0  p.Leu747_Pro753delinsSer  p.L747_P753delinsS   \n",
       "1               p.His271Arg             p.H271R   \n",
       "2        p.Tyr519TrpfsTer96        p.Y519Wfs*96   \n",
       "3               p.Tyr205Cys             p.Y205C   \n",
       "4               p.Glu294Ter             p.E294*   \n",
       "\n",
       "                                    Codons tumor_vaf normal_vaf  \\\n",
       "0                tTAAGAGAAGCAACATCTCcg/tcg  0.474465        0.0   \n",
       "1                                  cAc/cGc  0.186813        0.0   \n",
       "2  tACTACAATGCTCAGAGCCAGCAGTACCTGTACTGg/tg  0.442478        0.0   \n",
       "3                                  tAt/tGt  0.372483        0.0   \n",
       "4                                  Gag/Tag  0.360231        0.0   \n",
       "\n",
       "                      Hugo_HGVSp  \n",
       "0  EGFR_p.Leu747_Pro753delinsSer  \n",
       "1             PDGFRB_p.His271Arg  \n",
       "2       RBM10_p.Tyr519TrpfsTer96  \n",
       "3               TP53_p.Tyr205Cys  \n",
       "4               TP53_p.Glu294Ter  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate VAF for tumor and normal samples\n",
    "mut_data['tumor_vaf'] = mut_data['t_alt_count'] / (mut_data['t_ref_count'] + mut_data['t_alt_count'])\n",
    "mut_data['normal_vaf'] = mut_data['n_alt_count'] / (mut_data['n_ref_count'] + mut_data['n_alt_count'])\n",
    "\n",
    "# Display the updated DataFrame\n",
    "mut_data.head()\n",
    "\n",
    "#these needs to be edited because VAF is different based on various mutation factors but the point stands that VAF aggregates a bunch of metrics into a single measurement that can be used to look at quantifiable mutation levels between samples\n",
    "\n",
    "# Create a new column 'Hugo_HGVSp' by appending 'Hugo_Symbol' and 'HGVSp'\n",
    "mut_data['Hugo_HGVSp'] = mut_data['Hugo_Symbol'] + '_' + mut_data['HGVSp'].astype(str)\n",
    "\n",
    "# Display the updated DataFrame\n",
    "mut_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Join le tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    Column  Unique Entries  Total Entries\n",
      "0              Hugo_Symbol             524         208953\n",
      "1           Entrez_Gene_Id             508         208953\n",
      "2               Chromosome              23         208953\n",
      "3              Consequence              83         208950\n",
      "4   Variant_Classification              20         208953\n",
      "5             Variant_Type               6         208953\n",
      "6         Reference_Allele            5795         208953\n",
      "7        Tumor_Seq_Allele1            5795         208953\n",
      "8        Tumor_Seq_Allele2            1305         208953\n",
      "9     Tumor_Sample_Barcode           23867         208953\n",
      "10         Mutation_Status               2         208953\n",
      "11                   Score               1         208953\n",
      "12             t_ref_count            2063         208953\n",
      "13             t_alt_count            1443         208953\n",
      "14             n_ref_count            1594         208953\n",
      "15             n_alt_count              48         208953\n",
      "16                   HGVSc          138579         208444\n",
      "17                   HGVSp           96828         200433\n",
      "18             HGVSp_Short           98438         208436\n",
      "19                  Codons            9562         200574\n",
      "20               tumor_vaf           91271         208953\n",
      "21              normal_vaf            4070         208846\n",
      "22              Hugo_HGVSp          131265         208953\n"
     ]
    }
   ],
   "source": [
    "#identifying a unique ID for columns\n",
    "# Print the number of entries and unique entries in each column of mut_data\n",
    "column_info = mut_data.nunique().reset_index()\n",
    "column_info.columns = ['Column', 'Unique Entries']\n",
    "column_info['Total Entries'] = mut_data.count().values\n",
    "print(column_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Duplicates in 'Hugo_HGVSp':\n",
      "                           Hugo_HGVSp Tumor_Sample_Barcode\n",
      "0       EGFR_p.Leu747_Pro753delinsSer    P-0081657-T01-IM7\n",
      "3                    TP53_p.Tyr205Cys    P-0083825-T01-IM7\n",
      "4                    TP53_p.Glu294Ter    P-0083825-T01-IM7\n",
      "6            FOXA1_p.Gly327ArgfsTer17    P-0083825-T01-IM7\n",
      "15             TP53_p.Asn239ThrfsTer8    P-0000106-T01-IM3\n",
      "...                               ...                  ...\n",
      "208946                       TP53_nan    P-0014611-T01-IM6\n",
      "208947                       TCF3_nan    P-0014611-T01-IM6\n",
      "208948              PTPRT_p.Arg183Leu    P-0014611-T01-IM6\n",
      "208950                       ATRX_nan    P-0014611-T01-IM6\n",
      "208951                BTK_p.Gly594Trp    P-0014611-T01-IM6\n",
      "\n",
      "[96601 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Find duplicates in the 'Hugo_HGVSp' column\n",
    "duplicates = mut_data[mut_data.duplicated('Hugo_HGVSp', keep=False)]\n",
    "print(\"Duplicates in 'Hugo_HGVSp':\")\n",
    "print(duplicates[['Hugo_HGVSp', 'Tumor_Sample_Barcode']])\n",
    "\n",
    "# Create the pivot table\n",
    "#pivot_table = mut_data.pivot(index='Hugo_HGVSp', columns='Tumor_Sample_Barcode', values='tumor_vaf')\n",
    "#pivot_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dictionary to map Tumor_Sample_Barcode to Cancer Type\n",
    "barcode_to_cancer_type = dict(zip(clinical_sample['#Sample Identifier'], clinical_sample['Cancer Type']))\n",
    "\n",
    "# Add a new column 'cancer_type' to mut_data based on the mapping\n",
    "mut_data['cancer_type'] = mut_data['Tumor_Sample_Barcode'].map(barcode_to_cancer_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
